Source/Core/InputCommon/InputProfile.cpp

```diff
@@ -0,0 +1,187 @@
+    return Common::DoFileSearch({ device_profile_root_location }, { ".ini" });
+  }
+
+  std::string GetProfile(CycleDirection cycle_direction, int& profile_index, const std::vector<std::string>& profiles)
+  {
+    // update the index and bind it to the number of available strings
+    auto positive_modulo = [](int& i, int n) {i = (i % n + n) % n;};
+    profile_index += static_cast<int>(cycle_direction);
+    positive_modulo(profile_index, static_cast<int>(profiles.size()));
+
+    return profiles[profile_index];
+  }
+
+  void UpdateProfile(const std::string& profile_filename, const std::vector<ControllerEmu::EmulatedController*> controllers)
+  {
+    std::string base;
+    SplitPath(profile_filename, nullptr, &base, nullptr);
+
+    IniFile ini_file;
+    if (ini_file.Load(profile_filename))
+    {
+      Core::DisplayMessage("Loading input profile: " + base, s_display_message_ms);
+
+      for (auto* controller : controllers)
+      {
+        controller->LoadConfig(ini_file.GetOrCreateSection("Profile"));
+        controller->UpdateReferences(g_controller_interface);
+      }
+    }
+    else
+    {
+      Core::DisplayMessage("Unable to load input profile: " + base, s_display_message_ms);
+    }
+  }
+
+  std::vector<ControllerEmu::EmulatedController*> GetControllersForDevice(InputConfig* device_configuration)
+  {
+    std::size_t size = device_configuration->GetControllerCount();
+
+    std::vector<ControllerEmu::EmulatedController*> result;
+    result.reserve(size);
+
+    for (int i = 0; i < static_cast<int>(size); i++)
+    {
+      result.push_back(device_configuration->GetController(i));
+    }
+
+    return result;
+  }
+
+  std::vector<std::string> GetProfilesFromString(const std::string& setting, InputConfig* device_configuration)
+  {
+    const auto& profiles = SplitString(setting, ',');
+
+    const std::string device_profile_root_location(File::GetUserPath(D_CONFIG_IDX) + "Profiles/" + device_configuration->GetProfileName());
+
+    std::vector<std::string> result;
+    result.resize(profiles.size());
+    std::transform(profiles.begin(), profiles.end(), result.begin(), [&device_profile_root_location](const std::string& profile)
+    {
+      return device_profile_root_location + "/" + profile;
+    });
+
+    return result;
+  }
+
+  std::vector<std::string> GetMatchingProfilesFromSetting(const std::string& setting, const std::vector<std::string> profiles, InputConfig* device_configuration)
+  {
+    const auto& profiles_from_setting = GetProfilesFromString(setting, device_configuration);
+    if (profiles_from_setting.empty())
+    {
+      return {};
+    }
+
+    std::vector<std::string> result;
+    std::set_intersection(profiles.begin(), profiles.end(), profiles_from_setting.begin(), profiles_from_setting.end(), back_inserter(result));
+    return result;
+  }
+}
+
+namespace InputProfile
+{
+  void NextWiimoteProfile()
```

Comment:
There seems like quite an awful lot of duplication going on here for what is essentially just a simple change in direction. Perhaps consider factoring this out in such a way where it's minimized more. (e.g. an implemetation function that is just called from both prev and next variants, which only serve to provide the direction as an argument to that implementation function).

Reply:
Yeah, it did come to mind but I wanted to get all my requirements met before going back and fleshing out the code design.

I'll add this to my to-do list.  Thanks @lioncash !